[FrameworkBundle] Add wiring for cache-pool and cache-adapter service tags

nicolas-grekas opened the pull request:
This is a pretty heavy change for this feature :) I don't care for the "commit" credit so if you agree, please squash this into your own branch.

I propose to split two concepts:
- cache adapters
- and cache pools.

Cache pools are what user manipulates: each pool service has access to its own set of cache items.
Cache pool use cache adapters for the implementations:
Cache adapters are abstract services that have the cache-adapter tag, themselves having the namespace-arg-index attribute.

This is the most flexible way of wiring cache pools I can imagine: it allows everyone to define cache.pool and share state between several services.
With the framework bundle configuration, it's even easier as the basic adapters are preconfgured.

nicolas-grekas committed:
[FrameworkBundle] Add wiring for cache-pool and cache-adapter service tags

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -555,40 +555,22 @@ private function addCacheSection(ArrayNodeDefinition $rootNode)
-                                        }
-
-                                        foreach ($v as $key => $value) {
-                                            if (!in_array($key, array('type', 'name', 'options'))) {
```
I don't get why we need to move options into the "options" subkey. Let's remove this if it's not required.

nicolas-grekas commented:
Note that I did not run the code yet, nor did I update the tests. If you like it, don't hesitate to take over :)

xabbuh commented:
@nicolas-grekas I just have one question. :) How would I as a user now being able to simply inject a pool into my services?

With how the code looked like before you could have used something like this:

``` yaml
my_service:
    class: Foo
    arguments:
        - '@cache.adapter.bar'
```

Where the `bar` adapter was configured like this:

``` yaml
framework:
    cache:
        bar:
            type: filesystem
            options:
                directory: /tmp
```

nicolas-grekas commented:
You'd do the same but the service would be called cache.pool.bar

nicolas-grekas commented:
For the config, I removed the "options" nesting level, and added a default to the directory setting in FilesystemAdapter. Which default to kernel.cache_dir in the FwExt

nicolas-grekas commented:
~~BTW, I just thought that we could replace the namespace_arg_index tag attribute by a %cache_pool_namespace% parameter that would be resolved in the compiler pass. We could do the same for all the other config options: take them from the config, put them as attributes on the cache.pool tags, then apply them to cache adapters using parameter resolution in the compiler pass. That would allow to easily configure nested adapters injection (like chained adapters or proxyfied adapters)!~~

too complex :)

nicolas-grekas commented:
Includes part of https://github.com/symfony/symfony/pull/18024. Once it is merged, we should rebase and add a "proxy" cache type IMHO.

xabbuh commented:
Of course, the `cache.pool.` services are not abstract so they can be injected as is. But what would I do if I wanted to configure only one pool, inject it everywhere, but I want the namespace prefix to be different everywhere? I now have to configure one pool for each of them, haven't I?

nicolas-grekas commented:
You're right. And this collides with the use case of injecting the same pool to several services (e.g. a cache warmer and a cache consumer). What should we do?

xabbuh commented:
What about if we introduced a combination of both my initial idea and the changes you did here? So basically you could configure cache adapters only, inject them and a pool would be created for you automatically having its own namespace while configuring a pool would make it possible to reuse the same pool in several services?

nicolas-grekas commented:
Looks like a good idea! Is it possible to not use the prefixing convention but instead rely solely on the cache.adapter tag to do this ?

xabbuh commented:
You mean creating the pool on demand when a service is injected somewhere that is tagged with the `cache.adapter` tag?

xabbuh commented:
By the way, maybe we should simply add a `prototype` option to the `cache.pool` config options which could be set to `true` to trigger that the service will be tagged with `cache.adapter`.

nicolas-grekas commented:
The more I think about this the more I'm confused :) Do we really need to auto-create a separate cache pool for each use? I fear this may become too magic for users. Maybe we should get back the discussion on the original thread and see what others think?

xabbuh commented:
Yes, makes sense. I'll integrate your changes later today and fix the tests. We can then continue the discussion back in Symfony repo.

nicolas-grekas committed:
[FrameworkBundle] Add PSR-6 cache pool proxying

nicolas-grekas commented:
PR rebased and ProxyAdapter added to the allowed configs

xabbuh commented:
I pushed your changes as well as a few fixes and tests. I will write a comment later today to sum of for everyone where we are no and how we could proceed.
